app/controllers/amr_heatmap_controller.rb

```diff
@@ -58,4 +63,167 @@ def amr_counts
 
     render json: amr_data
   end
+
+  def fetch_card_info
+    gene_name = params[:geneName]
+    ontology = {
+      "accession" => "",
+      "label" => "",
+      "synonyms" => "",
+      "description" => "",
+      "geneFamily" => "",
+      "drugClass" => "",
+      "publications" => [],
+      "error" => ""
+    }
+    card_owl = get_s3_file(S3_CARD_OWL)
+    if card_owl.nil?
+      ontology["error"] = "Unable to retrieve CARD ARO owl file."
+      render json: ontology
+      return
+    end
+    ## Search the CARD OWL file
+    raw_owl_xml = card_owl.force_encoding("utf-8")
+    owl_doc = Nokogiri::XML(raw_owl_xml)
+    search_result = search_card_owl(gene_name, owl_doc)
+    unless search_result["error"].nil?
+      ontology["error"] = search_result["error"]
+      render json: ontology
+      return
+    end
+
+    ontology_info = gather_ontology_information(search_result["matching_node"], owl_doc)
+    ontology_info.each do |key, value|
+      ontology[key] = value
+    end
+
+    ontology["publications"] = ontology["publications"].map do |pubmed_string|
+      get_publication_name(pubmed_string.split(":")[1])
+    end
+
+    render json: ontology
+  end
+
+  private
+
+  def search_card_owl(gene_name, xml_dom)
+    search_result = {}
+    owl_classes = xml_dom.xpath(".//owl:Class")
+    matching_entry = nil
+
+    # normalize gene name
+    alphanumeric_gene_name = gene_name.downcase.gsub(/\W/, '')
+    regex_for_gene = Regexp.new("\\b#{alphanumeric_gene_name}\\b")
+    # search through owl classes
+    owl_classes.each do |owl_class|
+      # first we check the label
+      label = owl_class.at_xpath("./rdfs:label").content
+      separated_label = label.downcase.tr('/', ' ')
+      alphanumeric_label = separated_label.gsub(/[^0-9a-z_\s]/, '')
+      if regex_for_gene.match?(alphanumeric_label)
+        matching_entry = owl_class
+        break
+      end
+
+      # then we check the synonyms
+      synonyms = owl_class.xpath("./oboInOwl:hasExactSynonym")
+      synonyms.each do |synonym|
+        separated_synonym = synonym.content.downcase.tr('/', ' ')
+        alphanumeric_synonym = separated_synonym.gsub(/[^0-9a-z_\s]/, '')
+        if regex_for_gene.match?(alphanumeric_synonym)
+          matching_entry = owl_class
+          break
+        end
+      end
+    end
+
+    # abort if no matching entry
+    if matching_entry.nil?
+      search_result["error"] = "No match found for #{gene_name} in the CARD Antibiotic Resistance Ontology."
```

Comment:
you can put the error message in a template constant like `err = "No match found for %s in the CARD Antibiotic Resistance Ontology."` and then `err % [gene_name]`